keymanapp/keyman
# fix(web): inferred key text + spacebar caption scaling

Fixes #7615.

Also fixes a highly-related issue with spacebar captions; I've noted that the text will often overflow in embedded contexts.  It's not a perfect fix, as there appears to be a minimum text size allowed by browsers, but it's definitely an improvement for affected cases.  `gff_amharic` makes a fantastic example for this due to its relatively narrow spacebar; I'd noticed this tidbit while working on other recent issues.

With the fix:

![image](https://user-images.githubusercontent.com/25213402/202398602-220cb6c6-360b-418f-a9cd-5c0b89a207f2.png)

Without it:

![image](https://user-images.githubusercontent.com/25213402/202398686-28d437fd-6bc9-4761-91aa-581c1232559d.png)

----

Meanwhile, when there is sufficient space:

With the fix:
![image](https://user-images.githubusercontent.com/25213402/202399196-57f4df98-c822-4693-939c-8efba525419f.png)

Without it:
![image](https://user-images.githubusercontent.com/25213402/202399376-36912020-41f3-41fe-80dc-c42c61800405.png)

Here, the padding-maintenance logic kicked in and shrunk this case a bit more than necessary, but it at least demonstrates the level of effect to expect.

----

Finally, for an extremely contrived but obvious example...

![image](https://user-images.githubusercontent.com/25213402/202399853-95161a65-270d-4e75-9355-0caf2ce35786.png)

Note that I've retooled the old test page (for the  _highly_ related #382, which kicked off the key-cap scaling feature) and its keyboard to better test #7615.

----

## User Testing

- **TEST_KEYCAP_SCALING**
    1. Using Chrome's remote-device emulation... 
    2. Load Web's "Test page for automatic key-cap scaling".
    3. Click the "Type here" element to display the keyboard.
    4. Verify that the first two keys between the "g" and "l" keys appear identical, with both displaying the same character 5 times.
    5. Verify that the full word "Keyman" is visible on the key immediately left of the "l".

- **TEST_SPACEBAR_SCALING**
    1. Using Chrome's remote-device emulation...
    2. Load Web's "Test page for automatic key-cap scaling".
    3. Click the "Type here" element to display the keyboard.
    4. Verify that no text runs off either edge of the spacebar.
       - The full text: `English - Keycap Scaling Test with a really long name in order to also test captions`

- **TEST_IN_APP_SPACEBAR_SCALING**
    1. Using the Android app on a **tablet** form-factor device...
    2. Go to Settings => Spacebar caption and select "Language + Keyboard".
    3. Using Settings => Install Keyboard or Dictionary => Install from keyman.com, install `gff_amharic`.
    4. Verify that no text runs off either edge of the spacebar.
